### ISSUE TITLE
Spring boot 2.0.1 - Actuator endpoint failing when content negotiation`favorPathExtension` set as true

### ISSUE THREAD

**Issue by JagmohanSharma** [spring-projects/spring-boot#12900]
Getting error of HTTP ERROR 406 Not acceptable content, when using spring boot 2.0.1.

when we set favorPathExtension to true , then we get HTTP ERROR 406 Not acceptable content for each actuator endpoint.

but it was working fine with spring boot 2.0.0.

As now in spring boot 2.0.1, we have defined order for `WebMvcAutoConfigurationAdapter` to be executed at higher precedence so that any custom configuration get executed after this.

But it is causing issue for actuator endpoint when we set `favorPathExtension` as true.

**Comment by JagmohanSharma** [comment]
When we set `favorPathExtension ` to true , only 'HeaderContentNegotiationStrategy' is getting executed but not `PathExtensionContentNegotiationStrategy` as per `favorPathExtension ` value as true. So we get 406 error.

**Comment by philwebb** [comment]
@JagmohanSharma Is this for any actuator endpoint or some custom ones that you're developing? Do you have a sample application that we can run for ourselves to see what you mean?

**Comment by JagmohanSharma** [comment]
This is for actuator endpoints. Please find sample code here https://github.com/JagmohanSharma/SpringBoot2Issues. I added custom content negotiation configuration but disabled it. Now it is a plain spring boot configured and have set `spring.mvc.contentnegotiation.favor-path-extension=true` with spring boot 2.0.1. So here test `ContentNegotiationTest` is getting failed where I am executing info endpoint.

This works fine if we use spring boot 2.0.0 OR set `spring.mvc.contentnegotiation.favor-path-extension=false` with 2.0.1 itself.

**Comment by JagmohanSharma** [comment]
I found some more info regarding this issue.

In `OptionalPathExtensionContentNegotiationStrategy`, we have below method which returns empty list if `SKIP_ATTRIBUTE` is set as true or not set.

```
@Override
		public List<MediaType> resolveMediaTypes(NativeWebRequest webRequest)
				throws HttpMediaTypeNotAcceptableException {
			Object skip = webRequest.getAttribute(SKIP_ATTRIBUTE,
					RequestAttributes.SCOPE_REQUEST);
			if (skip != null && Boolean.parseBoolean(skip.toString())) {
				return Collections.emptyList();
			}
			return this.delegate.resolveMediaTypes(webRequest);
		}
```

Now in `ContentNegotiationManager`, if we get empty list then it simply returns that and any other strategy will not be executed. 

```
@Override
	public List<MediaType> resolveMediaTypes(NativeWebRequest request) throws HttpMediaTypeNotAcceptableException {
		for (ContentNegotiationStrategy strategy : this.strategies) {
			List<MediaType> mediaTypes = strategy.resolveMediaTypes(request);
			if (mediaTypes.equals(MEDIA_TYPE_ALL_LIST)) {
				continue;
			}
			return mediaTypes;
		}
		return MEDIA_TYPE_ALL_LIST;
	}
```

this is called from `AbstractMessageConverterMethodProcessor` while executing below code for getting AcceptableMediaTypes which we get as empty list from `ContentNegotiationManager`

```
else {
			HttpServletRequest request = inputMessage.getServletRequest();
			List<MediaType> requestedMediaTypes = getAcceptableMediaTypes(request);
```

SO we get exception while comparing this acceptable media type with producible media types in `AbstractMessageConverterMethodProcessor`

```
HttpServletRequest request = inputMessage.getServletRequest();
			List<MediaType> requestedMediaTypes = getAcceptableMediaTypes(request);
			List<MediaType> producibleMediaTypes = getProducibleMediaTypes(request, valueType, declaredType);

			if (outputValue != null && producibleMediaTypes.isEmpty()) {
				throw new HttpMessageNotWritableException(
						"No converter found for return value of type: " + valueType);
			}
			mediaTypesToUse = new ArrayList<>();
			for (MediaType requestedType : requestedMediaTypes) {
				for (MediaType producibleType : producibleMediaTypes) {
					if (requestedType.isCompatibleWith(producibleType)) {
						mediaTypesToUse.add(getMostSpecificMediaType(requestedType, producibleType));
					}
				}
			}
			if (mediaTypesToUse.isEmpty()) {
				if (outputValue != null) {
					throw new HttpMediaTypeNotAcceptableException(producibleMediaTypes);
				}
				return;
			}
```

**Comment by bclozel** [comment]
Thanks @JagmohanSharma for the report and the detailed analysis. This fix should be available shortly in Spring Boot 2.0.2.BUILD-SNAPSHOT and 2.1.0.BUILD-SNAPSHOT. Let us know if you manage to validate this fix.
